File: StepMotor/app/src/main/java/io/singularity/stepmotor/MainActivity.java
package io.singularity.stepmotor;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;

import com.google.android.things.pio.Gpio;
import com.google.android.things.pio.PeripheralManagerService;
import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.FirebaseDatabase;

import java.io.IOException;

public class MainActivity extends AppCompatActivity implements ChildEventListener {
    private static final String TAG = MainActivity.class.getSimpleName();
    private FirebaseDatabase firebaseDatabase;
    private Gpio gpio1;
    private Gpio gpio2;
    private Gpio gpio3;
    private Gpio gpio4;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        PeripheralManagerService service = new PeripheralManagerService();

        try {
            gpio1 = service.openGpio("BCM6");
            gpio1.setDirection(Gpio.DIRECTION_OUT_INITIALLY_LOW);

            gpio2 = service.openGpio("BCM13");
            gpio2.setDirection(Gpio.DIRECTION_OUT_INITIALLY_LOW);

            gpio3 = service.openGpio("BCM19");
            gpio3.setDirection(Gpio.DIRECTION_OUT_INITIALLY_LOW);

            gpio4 = service.openGpio("BCM26");
            gpio4.setDirection(Gpio.DIRECTION_OUT_INITIALLY_LOW);

        } catch (IOException e) {
            Log.e(TAG, "Error on PeripheralIO API", e);
        } catch (Exception e){
            e.printStackTrace();
        }

        firebaseDatabase = FirebaseDatabase.getInstance();
        firebaseDatabase.getReference("stepMotor/").addChildEventListener(this);
    }

    @Override
    public void onChildAdded(DataSnapshot dataSnapshot, String s) {

    }

    @Override
    public void onChildChanged(DataSnapshot dataSnapshot, String s) {
        Log.d("Test", "onChildChanged : " + dataSnapshot.getValue());

        String isEnable = (String) dataSnapshot.getValue();
        for (int i=0; i<1000; i++){
            setLedEnable(Integer.parseInt(isEnable));
            setLedEnable(-1);
        }
    }

    @Override
    public void onChildRemoved(DataSnapshot dataSnapshot) {

    }

    @Override
    public void onChildMoved(DataSnapshot dataSnapshot, String s) {

    }

    @Override
    public void onCancelled(DatabaseError databaseError) {

    }

    public void setLedEnable(int ledEnable) {
        try {
            switch (ledEnable) {
                case 1:
                    gpio1.setValue(true);
                    gpio4.setValue(true);
                    break;
                case 2:
                    gpio3.setValue(true);
                    break;
                case 3:
                    gpio2.setValue(true);
                    break;
                case 4:
                    gpio2.setValue(true);
                    gpio3.setValue(true);
                    break;
                case -1:
                    gpio1.setValue(false);
                    gpio2.setValue(false);
                    gpio3.setValue(false);
                    gpio4.setValue(false);
            }
        } catch (IOException e) {
            Log.e(TAG, e.getMessage());
        }

        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

    }
}

File: FirstAnroidThingsWithFirebase/app/src/main/java/com/akj/firstandroidthings/MainActivity.kt
package com.akj.firstandroidthings

import android.app.Activity
import android.os.Bundle
import android.util.Log
import com.google.android.things.pio.Gpio
import com.google.android.things.pio.PeripheralManagerService
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import java.io.IOException
import java.util.*
import kotlin.concurrent.schedule

class MainActivity : Activity() {

    val TAG = "MainActivity"

    /**
     * GPIO(General-purpose input/output) 의 이름.
     */
    val gpioName = "BCM6"

    /**
     * GPIO 인스턴스
     */
    var gpio: Gpio? = null

    /**
     * Firebase Reference
     */
    val database = FirebaseDatabase.getInstance().reference

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // PeripheralManagerService 는 주변장치 관리자 같은거
        val service = PeripheralManagerService()

        // gpio 현재 하드웨어에 존재하는 GPIO 리스트를 로깅
        Log.d(TAG, "Available GPIO: " + service.gpioList)

        try {
            gpio = service.openGpio(gpioName)

            // 입출력 설정을 체크한다.
            // DIRECTION_OUT_INITIALLY_LOW 는 GPIO 를 출력모드로 설정
            gpio?.setDirection(Gpio.DIRECTION_OUT_INITIALLY_LOW);

            gpio?.let { gpio ->
                // Firebase 에서 bcm6 경로를 읽어온다.
                database.child("bcm6").addValueEventListener(object:ValueEventListener{
                    override fun onCancelled(error: DatabaseError?) {
                    }

                    override fun onDataChange(snapshot: DataSnapshot?) {
                        // Firebase 의 bcm6 데이터가 on 이면 LED 를 켜고 그렇지 않으면 끈다.
                        gpio.value = "on".equals(snapshot?.value.toString())
                    }
                })
            }

        } catch (e: IOException) {
            Log.e(TAG, "Error on PeripheralIO API", e);
        }
    }
}
